File: packages/memory/src/mcp-client.ts
import { GeneratedMcpToolsClient } from './generated/mcp/methods.js'

export interface McpToolsClientOptions {
  apiKey: string
  baseUrl?: string
  fetch?: typeof globalThis.fetch
  listRetries?: number
  retryDelayMs?: number
  /** Bounded deadline for each MCP HTTP call. Defaults to 300 seconds. */
  requestTimeoutMs?: number
}

export interface McpToolDescriptor {
  name: string
  title?: string
  description?: string
  inputSchema: Record<string, unknown>
  outputSchema?: Record<string, unknown>
  annotations?: Record<string, unknown>
}

export interface McpToolErrorDetails {
  httpStatus?: number
  rpcCode?: number
  data?: unknown
  toolError?: unknown
}

export class McpToolError extends Error {
  readonly httpStatus?: number
  readonly rpcCode?: number
  readonly data?: unknown
  readonly toolError?: unknown

  constructor(message: string, details: McpToolErrorDetails = {}) {
    super(message)
    this.name = 'McpToolError'
    this.httpStatus = details.httpStatus
    this.rpcCode = details.rpcCode
    this.data = details.data
    this.toolError = details.toolError
  }
}

interface JsonRpcErrorPayload {
  code: number
  message: string
  data?: unknown
}

export interface McpContentBlock {
  type: string
  text?: string
  data?: string
  mimeType?: string
  uri?: string
  name?: string
  description?: string
  resource?: unknown
  [key: string]: unknown
}

export interface McpToolCallResult {
  content?: McpContentBlock[]
  structuredContent?: unknown
  isError?: boolean
}

interface JsonRpcResponse<T> {
  error?: JsonRpcErrorPayload
  result?: T
}

class McpJsonRpcTransport {
  private readonly apiKey: string
  private readonly baseUrl: string
  private readonly fetchImpl: typeof globalThis.fetch
  private readonly listRetries: number
  private readonly retryDelayMs: number
  private readonly requestTimeoutMs: number
  private rpcId = 0

  constructor(options: McpToolsClientOptions) {
    this.apiKey = options.apiKey
    this.baseUrl = (options.baseUrl ?? 'https://mcpscraper.dev').replace(/\/$/, '')
    this.fetchImpl = options.fetch ?? globalThis.fetch
    this.listRetries = options.listRetries ?? 2
    this.retryDelayMs = options.retryDelayMs ?? 250
    this.requestTimeoutMs = options.requestTimeoutMs ?? 300_000
    if (!Number.isFinite(this.requestTimeoutMs) || this.requestTimeoutMs <= 0) throw new TypeError('requestTimeoutMs must be a positive number')
  }

  private async request<T>(method: string, params?: Record<string, unknown>): Promise<T> {
    const attempts = method === 'tools/list' ? this.listRetries + 1 : 1
    let response: Response | undefined
    for (let attempt = 0; attempt < attempts; attempt += 1) {
      try {
        response = await this.fetchImpl(`${this.baseUrl}/mcp`, {
          method: 'POST',
          headers: {
            'x-api-key': this.apiKey,
            'content-type': 'application/json',
            accept: 'application/json, text/event-stream',
          },
          body: JSON.stringify({
            jsonrpc: '2.0',
            id: ++this.rpcId,
            method,
            ...(params ? { params } : {}),
          }),
          signal: AbortSignal.timeout(this.requestTimeoutMs),
        })
      } catch (error) {
        if (error instanceof DOMException && (error.name === 'TimeoutError' || error.name === 'AbortError')) {
          const envelope = {
            error_code: 'mcp_request_timeout',
            error_type: 'timeout',
            message: `MCP request exceeded ${Math.round(this.requestTimeoutMs / 1000)}s and was cancelled client-side. The operation may still be running; poll its durable job or retry with the same idempotency key.`,
            retryable: true,
          }
          throw new McpToolError(envelope.message, { httpStatus: 408, toolError: envelope })
        }
        throw error
      }
      const transientDiscoveryFailure = method === 'tools/list' && [502, 503, 504].includes(response.status)
      if (!transientDiscoveryFailure || attempt === attempts - 1) break
      await new Promise(resolve => setTimeout(resolve, this.retryDelayMs * (attempt + 1)))
    }

    if (!response) throw new McpToolError(`MCP method "${method}" produced no HTTP response`)

    if (!response.ok) {
      const raw = await response.text().catch(() => '')
      let data: unknown = raw
      try { data = JSON.parse(raw) } catch { /** keep text */ }
      const publicBody = data && typeof data === 'object' && typeof (data as Record<string, unknown>).error_code === 'string'
        ? data as Record<string, unknown>
        : null
      const message = publicBody && typeof publicBody.message === 'string'
        ? publicBody.message
        : `HTTP ${response.status} calling MCP method "${method}"`
      throw new McpToolError(message, {
        httpStatus: response.status,
        data,
        ...(publicBody ? { toolError: publicBody } : {}),
      })
    }

    const payload = (await response.json()) as JsonRpcResponse<T>
    if (payload.error) {
      throw new McpToolError(payload.error.message, {
        rpcCode: payload.error.code,
        data: payload.error.data,
      })
    }
    if (payload.result === undefined) throw new McpToolError(`MCP method "${method}" returned no result`)
    return payload.result
  }

  async listTools(): Promise<McpToolDescriptor[]> {
    const result = await this.request<{ tools?: McpToolDescriptor[] }>('tools/list')
    return result.tools ?? []
  }

  async callToolResult(name: string, args: unknown = {}): Promise<McpToolCallResult> {
    const result = await this.request<McpToolCallResult>('tools/call', {
      name,
      arguments: (args ?? {}) as Record<string, unknown>,
    })
    if (result.isError) {
      const parsed = this.parsedToolValue(result)
      const message =
        parsed && typeof parsed === 'object' && 'message' in (parsed as Record<string, unknown>)
          ? String((parsed as Record<string, unknown>).message)
          : typeof parsed === 'string'
            ? parsed
            : `Tool "${name}" failed`
      throw new McpToolError(message, { toolError: parsed })
    }
    return result
  }

  private parsedToolValue(result: McpToolCallResult): unknown {
    const textBlock = result.content?.find(block => block.type === 'text')?.text
    let parsed: unknown = textBlock
    if (typeof textBlock === 'string') {
      try {
        parsed = JSON.parse(textBlock)
      } catch {
        parsed = textBlock
      }
    }
    if (result.structuredContent !== undefined) parsed = result.structuredContent
    return parsed
  }

  async callTool(name: string, args: unknown = {}): Promise<unknown> {
    return this.parsedToolValue(await this.callToolResult(name, args))
  }
}

export class McpToolsClient extends GeneratedMcpToolsClient {
  private readonly transport: McpJsonRpcTransport

  constructor(options: McpToolsClientOptions) {
    const transport = new McpJsonRpcTransport(options)
    super(transport.callTool.bind(transport))
    this.transport = transport
  }

  listTools(): Promise<McpToolDescriptor[]> {
    return this.transport.listTools()
  }

  async describeTool(name: string): Promise<McpToolDescriptor | undefined> {
    const tools = await this.listTools()
    return tools.find(tool => tool.name === name)
  }

  callTool(name: string, args: unknown = {}): Promise<unknown> {
    return this.transport.callTool(name, args)
  }

  callToolResult(name: string, args: unknown = {}): Promise<McpToolCallResult> {
    return this.transport.callToolResult(name, args)
  }
}

File: packages/memory-python/src/mcpscraper_memory/mcp_client.py
from __future__ import annotations

import itertools
import json
import time
from typing import Any

import requests

from ._mcp_generated_client import GeneratedMcpToolsClient


class McpToolError(RuntimeError):
    def __init__(
        self,
        message: str,
        *,
        http_status: int | None = None,
        rpc_code: int | None = None,
        data: Any = None,
        tool_error: Any = None,
    ) -> None:
        super().__init__(message)
        self.http_status = http_status
        self.rpc_code = rpc_code
        self.data = data
        self.tool_error = tool_error


class McpToolsClient(GeneratedMcpToolsClient):
    """Typed access to every tool exposed by https://mcpscraper.dev/mcp."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://mcpscraper.dev",
        session: requests.Session | None = None,
        list_retries: int = 2,
        retry_delay_s: float = 0.25,
    ) -> None:
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._session = session or requests.Session()
        self._rpc_id = itertools.count(1)
        self._list_retries = list_retries
        self._retry_delay_s = retry_delay_s
        super().__init__(self._call_tool)

    def _request(self, method: str, params: dict[str, Any] | None = None) -> Any:
        payload: dict[str, Any] = {
            "jsonrpc": "2.0",
            "id": next(self._rpc_id),
            "method": method,
        }
        if params is not None:
            payload["params"] = params
        attempts = self._list_retries + 1 if method == "tools/list" else 1
        response = None
        for attempt in range(attempts):
            response = self._session.post(
                f"{self._base_url}/mcp",
                headers={
                    "x-api-key": self._api_key,
                    "content-type": "application/json",
                    "accept": "application/json, text/event-stream",
                },
                json=payload,
            )
            transient_discovery_failure = method == "tools/list" and response.status_code in (502, 503, 504)
            if not transient_discovery_failure or attempt == attempts - 1:
                break
            time.sleep(self._retry_delay_s * (attempt + 1))
        assert response is not None
        if not response.ok:
            raise McpToolError(
                f'HTTP {response.status_code} calling MCP method "{method}"',
                http_status=response.status_code,
                data=response.text,
            )
        body = response.json()
        if body.get("error"):
            error = body["error"]
            raise McpToolError(
                error.get("message", f'MCP method "{method}" failed'),
                rpc_code=error.get("code"),
                data=error.get("data"),
            )
        if "result" not in body:
            raise McpToolError(f'MCP method "{method}" returned no result')
        return body["result"]

    def list_tools(self) -> list[dict[str, Any]]:
        result = self._request("tools/list")
        return result.get("tools", [])

    def describe_tool(self, name: str) -> dict[str, Any] | None:
        return next((tool for tool in self.list_tools() if tool.get("name") == name), None)

    def call_tool(self, name: str, args: dict[str, Any] | None = None) -> Any:
        return self._call_tool(name, args or {})

    @staticmethod
    def _parsed_tool_value(result: dict[str, Any]) -> Any:
        text_block = next(
            (block.get("text") for block in result.get("content", []) if block.get("type") == "text"),
            None,
        )
        parsed: Any = text_block
        if isinstance(text_block, str):
            try:
                parsed = json.loads(text_block)
            except ValueError:
                parsed = text_block
        if "structuredContent" in result:
            parsed = result["structuredContent"]
        return parsed

    def call_tool_result(self, name: str, args: dict[str, Any] | None = None) -> dict[str, Any]:
        """Return the complete MCP tool result, preserving image/audio/resource content blocks."""
        result = self._request("tools/call", {"name": name, "arguments": args or {}})
        parsed = self._parsed_tool_value(result)
        if result.get("isError"):
            message = (
                parsed.get("message")
                if isinstance(parsed, dict) and "message" in parsed
                else parsed
                if isinstance(parsed, str)
                else f'Tool "{name}" failed'
            )
            raise McpToolError(str(message), tool_error=parsed)
        return result

    def _call_tool(self, name: str, arguments: dict[str, Any]) -> Any:
        result = self.call_tool_result(name, arguments)
        parsed = self._parsed_tool_value(result)
        return parsed
